support/signal.c

```diff
@@ -212,42 +303,52 @@ Mono_Unix_UnixSignal_install (int sig)
 		if (sinfo.sa_handler != SIG_DFL || (void*)sinfo.sa_sigaction != (void*)SIG_DFL) {
 			pthread_mutex_unlock (&signals_mutex);
 			errno = EADDRINUSE;
-			return NULL;
+			return NULL; // This is an rt signal with an existing handler. Bail out.
 		}
 	}
 #endif /*defined (SIGRTMIN) && defined (SIGRTMAX)*/
 
+	// Scan through signals list looking for (1) an unused spot (2) a usable value for restore_handler
 	for (i = 0; i < NUM_SIGNALS; ++i) {
+		// We're still looking for a signal_info spot, and this one is available
+		int just_installed = 0;
 		if (h == NULL && signals [i].signum == 0) {
 			h = &signals [i];
-			h->handler = signal (sig, default_handler);
-			if (h->handler == SIG_ERR) {
-				h->handler = NULL;
+			h->restore_handler = signal (sig, default_handler);
+			if (h->restore_handler == SIG_ERR) {
+				h->restore_handler = NULL;
 				h = NULL;
 				break;
 			}
 			else {
-				h->have_handler = 1;
+				// Tenatively accept this restore handler
+				just_installed = 1;
 			}
 		}
-		if (!have_handler && signals [i].signum == sig &&
-				signals [i].handler != default_handler) {
-			have_handler = 1;
-			handler = signals [i].handler;
+		// Check if this slot has a "usable" (not installed by this file) restore handler
```

Comment:
What's a restore handler? We're just looking at the previous handler and being nice unix citizens by reinstalling the old one when we're done.